hikari/impl/rest.py

```diff
@@ -693,11 +703,11 @@ async def _request(
         while True:
             try:
                 uuid = time.uuid()
-                async with self._get_live_attributes().buckets.acquire(compiled_route):
+                async with live_attributes.still_alive().buckets.acquire(compiled_route):
```

Comment:
Tbh, this is still kinda meh. Would doing this in the  `__getattribute__` in live_attributes work?

```py
    def __getattribute__(self, attr: str) -> typing.Any:
        if super().__getattribute__("is_closing"):
            raise errors.ComponentStateConflictError("The REST client was closed mid-request")

        return super().__getattribute__(attr)

Reply:
sounds really implicit and bad for readability

Reply:
Having a `still_alive` check everywhere I find kind of unreadable (considering that _LiveAttributes could manage this with no need). A comment stating could be added on to specify this behavior

Reply:
> (considering that _LiveAttributes could manage this with no need)

Just because you can do something doesn't mean you should and are you really in-favour of stuff like having properties raise errors that aren't AttributeError, would you introduce that to the public api?

Reply:
> > (considering that _LiveAttributes could manage this with no need)
> 
> Just because you can do something doesn't mean you should and are you really in-favour of stuff like having properties raise errors that aren't AttributeError, would you introduce that to the public api?

No but, as I said, this is private detail, so we are kind of excused by doing something like this here. Another idea I just thought of is making properties for them and making the checks there (this should have been the first idea, sorry for that)

Reply:
> Another idea I just thought of is making properties for them and making the checks there (this should have been the first idea, sorry for that)

That's a lot of boilerplate and still as implicit and unreadable due to how it hides a function call which raises behind a property

Reply:
What about get methods that check if still alive and return the object?

```py
def get_buckets() -> ...:
   self._still_alive()
   return self._buckets
```

Reply:
how would you avoid making unnecessary calls to the check if it's already been verified that's it's active since the last yield to the event loop with that pattern

Reply:
Tbh, it is always better to be safe than sorry. But ig you are right.